java/src/main/native/src/TableJni.cpp

```diff
@@ -22,12 +22,185 @@
+    if (handle_buffer_method == nullptr) {
+      throw cudf::jni::jni_exception("handleBuffer method");
+    }
+
+    this->callback = env->NewGlobalRef(callback);
+    if (this->callback == nullptr) {
+      throw cudf::jni::jni_exception("global ref");
+    }
+  }
+
+  virtual ~jni_writer_data_sink() {
+    // This should normally be called by a JVM thread. If the JVM environment is missing then this
+    // is likely being triggered by the C++ runtime during shutdown. In that case the JVM may
+    // already be destroyed and this thread should not try to attach to get an environment.
+    JNIEnv* env = nullptr;
+    if (jvm->GetEnv(reinterpret_cast<void**>(&env), cudf::jni::MINIMUM_JNI_VERSION) == JNI_OK) {
+      env->DeleteGlobalRef(callback);
+      if (current_buffer != nullptr) {
+        env->DeleteGlobalRef(current_buffer);
+      }
+    }
+    callback = nullptr;
+    current_buffer = nullptr;
+  }
+
+  void host_write(void const* data, size_t size) override {
+    JNIEnv* env = cudf::jni::get_jni_env(jvm);
+    size_t left_to_copy = size;
+    const char * copy_from = static_cast<const char*>(data);
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy = left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char * copy_to = current_buffer_data + current_buffer_written;
+
+      std::memcpy(copy_to, copy_from, amount_to_copy);
+      copy_from = copy_from + amount_to_copy;
+      current_buffer_written += amount_to_copy;
+      total_written += amount_to_copy;
+      left_to_copy -= amount_to_copy;
+    }
+  }
+
+  bool supports_device_write() const override {
+    return true;
+  }
+
+  void device_write(void const* gpu_data, size_t size, cudaStream_t stream) {
+    if (this->stream != stream) {
+      if (is_stream_set) {
+        CUDA_TRY(cudaStreamSynchronize(this->stream));
+      }
+      this->stream = stream;
+    }
+    this->is_stream_set = true;
+
+    JNIEnv* env = cudf::jni::get_jni_env(jvm);
+    size_t left_to_copy = size;
+    const char * copy_from = static_cast<const char*>(gpu_data);
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy = left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char * copy_to = current_buffer_data + current_buffer_written;
+
+      CUDA_TRY(cudaMemcpyAsync(copy_to, copy_from, amount_to_copy, cudaMemcpyDeviceToHost, stream));
```

Comment:
Does the data sink API semantics guarantee it is safe to return before the data is copied?  In other words, we know that the entity calling the sink will synchronize on the stream passed to us after we return rather than implicitly relying on us to synchronize?

Reply:
This comment was not addressed.  cc: @nvdbaranec to see if this behavior is expected and documented in the API.